File: Python3/2961_double_modular_exponentiation.py
# https://leetcode.cn/problems/double-modular-exponentiation/

class Solution:
    def getGoodIndices(self, variables, target):
        """
        Inputs:
            variables: list[list[int]]
            target: int
        Outputs:
            res: list[int]
        """
        remainders = [[i] for i in range(10)]
        for i in range(10):
            val = i * i % 10
            while remainders[i][0] != val:
                remainders[i].append(val)
                val *= i
                val %= 10
        res = []
        for idx, (ai, bi, ci, mi) in enumerate(variables):
            if target >= mi:
                continue
            tmp = remainders[ai%10][bi % len(remainders[ai%10]) - 1]
            # Check if tmp ^ ci % mi == target.
            val, cnt = 1, 0
            while cnt < ci:
                val = val * tmp % mi
                cnt += 1
            if val % mi == target:
                res.append(idx)
        return res

File: Cpp/2961_double_modular_exponentiation.cpp
// https://leetcode.cn/problems/double-modular-exponentiation/

class Solution {
public:
    vector<int> getGoodIndices(vector<vector<int>>& variables, int target) {
        vector<vector<int>> remainders(10);
        for (int i = 0; i < 10; ++i) {
            remainders[i].push_back(i);
            int val = i * i % 10;
            while (val != i) {
                remainders[i].push_back(val);
                val = val * i % 10;
            }
        }
        vector<int> res;
        for (int i = 0; i < variables.size(); ++i) {
            int ai = variables[i][0], bi = variables[i][1], ci = variables[i][2];
            int mi = variables[i][3];
            if (mi <= target) continue;
            int tmp = remainders[ai%10][(bi-1) % remainders[ai%10].size()];
            // Check if tmp ^ ci % mi == target.
            int val = 1, cnt = 0;
            while (cnt < ci) {
                val = val * tmp % mi;
                cnt += 1;
            }
            if (val % mi == target) {
                res.push_back(i);
            }
        }
        return res;
    }
};
